File: Question/39_Combination_Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> cur;
        dfs(candidates, target, 0, 0, cur, ans);
        return ans;
    }
    void dfs(vector<int>& candidates, int target, int loc, int cur_sum, vector<int>& cur, vector<vector<int>>& ans) {
        if(cur_sum > target) return;
        if(loc >= candidates.size()) return;
        if(cur_sum == target) {
            ans.push_back(cur);
        }
        for(int i = loc; i < candidates.size(); ++i) {
            cur.push_back(candidates[i]);
            dfs(candidates, target, i, cur_sum + candidates[i], cur, ans);
            cur.pop_back();
        }
    }
};

File: Question_Java/39_Combination_Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        dfs(candidates, 0, target, new ArrayList<>(), ans);
        return ans;
    }

    private void dfs(int[] candidates, int loc, int target, List<Integer> curAns, List<List<Integer>> ans) {
        if (target == 0) {
            ans.add(new ArrayList<Integer>(curAns));
            return;
        }
        if (target < 0)
            return;

        for (int i = loc; i < candidates.length; ++i) {
            curAns.add(candidates[i]);
            dfs(candidates, i, target - candidates[i], curAns, ans);
            curAns.remove(curAns.size() - 1);
        }
    }
}